File: src/mrl/creatures/Human.java
/*
 *  "MysteriesRL", Java Roguelike game.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih (aka Alchemist, aka Norseman).
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mrl.creatures;

import java.awt.Color;
import java.util.ArrayList;
import java.util.List;
import jzrlib.common.CreatureSex;
import jzrlib.core.GameSpace;
import jzrlib.core.IEquippedCreature;
import jzrlib.core.body.AbstractBody;
import mrl.core.Locale;
import mrl.core.RS;
import mrl.creatures.body.HumanBody;
import mrl.game.events.CriminalEvent;
import mrl.game.events.SuspiciousSoundEvent;
import mrl.game.social.CrimeRecord;
import mrl.items.Item;
import mrl.items.ItemsList;
import mrl.maps.buildings.Building;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public class Human extends Creature implements IEquippedCreature
{
    private int fAge = 30;
    private Building fApartment;
    private String fName;
    private CreatureSex fSex = CreatureSex.csMale;

    private AbstractBody fBody;
    private final NPCStats fStats;
    private final ItemsList fInventory;
    private final Personality fPersonality;

    // <editor-fold defaultstate="collapsed" desc="sample">
    
    // </editor-fold>
    
    // <editor-fold defaultstate="collapsed" desc="Social stuff">
    
    private Human fParent;
    private Human fSpouse;
    private final List<Human> fChildrens;
    private final List<Human> fSiblings;
    public List<CrimeRecord> fCrimeRecords;
    
    // </editor-fold>

    public Human(GameSpace space, Object owner)
    {
        super(space, owner);

        this.fBody = new HumanBody(this);
        this.fStats = new NPCStats();
        this.fInventory = new ItemsList(this, true);
        this.fPersonality = new Personality();

        this.fParent = null;
        this.fSpouse = null;
        this.fChildrens = new ArrayList<>();
        this.fSiblings = new ArrayList<>();
        this.fCrimeRecords = new ArrayList<>();
        
        this.setHP(this.getHPMax());
    }

    @Override
    public char getAppearance()
    {
        if (this.isPlayer()) {
            return '@';
        }
        
        if (this.isAlive()) {
            switch (this.fSex) {
                case csFemale:
                    if (this.isAdult()) {
                        return 'W';
                    } else {
                        return 'w';
                    }

                case csMale:
                    if (this.isAdult()) {
                        return 'M';
                    } else {
                        return 'm';
                    }

                default:
                    return 'H';
            }
        } else {
            return '%';
        }
    }

    @Override
    public Color getAppearanceColor()
    {
        if (this.isPlayer()) {
            return Color.yellow;
        }
        
        return Color.white;
    }

    @Override
    public String getName()
    {
        return this.fName;
    }

    @Override
    public void setName(String value)
    {
        this.fName = value;
    }

    public final Building getApartment()
    {
        return this.fApartment;
    }

    public final void setApartment(Building apt)
    {
        this.fApartment = apt;
    }

    public final NPCStats getStats()
    {
        return fStats;
    }
    
    public final Personality getPersonality()
    {
        return this.fPersonality;
    }

    public final AbstractBody getBody()
    {
        return this.fBody;
    }

    public final void setBody(AbstractBody value)
    {
        if (this.fBody != null) {
            this.fBody.dispose();
        }
        this.fBody = value;
    }

    @Override
    public String getDesc()
    {
        StringBuilder result = new StringBuilder();
        result.append(this.getName() + ". ");
        
        boolean isMale = (this.getSex() == CreatureSex.csMale);
        
        String age;
        if (isMale) {
            age = Locale.format(RS.rs_HeIsNYearsOld, this.getAge());
        } else {
            age = Locale.format(RS.rs_SheIsNYearsOld, this.getAge());
        }
        result.append(age);
        
        /*if (this.getSpouse() != null) {
            result.append("His/her spouse " + this.getSpouse().getName() + ". ");
        }

        if (this.getParent() != null) {
            result.append("His/her parent " + this.getParent().getName() + ". ");
        }*/

        int rsid;
        if (isMale) {
            rsid = RS.rs_HeHasXHairAndYEyes;
        } else {
            rsid = RS.rs_SheHasXHairAndYEyes;
        }
        String pers = Locale.format(rsid, this.fPersonality.hairColor.name().toLowerCase(), this.fPersonality.eyeColor.name().toLowerCase());
        result.append(pers);
        
        return result.toString();
    }
    
    public final ItemsList getInventory()
    {
        return this.fInventory;
    }
    
    @Override
    public void doTurn()
    {
        HumanBody body = (HumanBody) this.fBody;
        body.update();

        if (!body.isStunned()){
            super.doTurn();
        }
    }

    @Override
    public void moveTo(int newX, int newY)
    {
        HumanBody body = (HumanBody) this.fBody;
        if (body.isFainted()) {
            return;
        }

        super.moveTo(newX, newY);

        body.adjustAttribute("stamina", -0.5f);
    }
    
    // <editor-fold defaultstate="collapsed" desc="Social stuff">
    
    public final int getAge()
    {
        return this.fAge;
    }

    public final void setAge(int age)
    {
        this.fAge = age;
    }

    public final CreatureSex getSex()
    {
        return this.fSex;
    }

    public final void setSex(CreatureSex sex)
    {
        this.fSex = sex;
    }

    public final boolean isAdult()
    {
        return this.fAge >= 18;
    }

    public final Human getParent()
    {
        return this.fParent;
    }

    public final void setParent(Human parent)
    {
        this.fParent = parent;
    }

    public final Human getSpouse()
    {
        return this.fSpouse;
    }

    public final void setSpouse(Human spouse)
    {
        if (this.fSpouse == null) {
            this.fSpouse = spouse;
            spouse.setSpouse(this);
        }
    }

    public final void addChild(Human child)
    {
        for (Human otherChild : this.fChildrens) {
            otherChild.addSibling(child);
        }

        this.fChildrens.add(child);
        child.setParent(this);
    }

    public final void addSibling(Human child)
    {
        if (!this.fSiblings.contains(child)) {
            this.fSiblings.add(child);
            child.addSibling(this);
        }
    }

    public final void addCrimeRecord(CrimeRecord crimeRecord)
    {
        for (CrimeRecord crime : fCrimeRecords) {
            if (crime.Type == crimeRecord.Type) {
                crime.incCount();
                return;
            }
        }
        fCrimeRecords.add(crimeRecord);
    }
    
    // </editor-fold>
    
    // <editor-fold defaultstate="collapsed" desc="Combat implementation">

    @Override
    public int getHPMax()
    {
        return this.fStats.getHPMax();
    }

    @Override
    public int getFovRadius()
    {
        int maxFov = (int) (7 + 1.2 * fStats.Per);
        int minFov = (int) (maxFov * 0.7);

        int fov = (int) (minFov + (maxFov - minFov) * this.getSpace().getLightFactor());

        return fov;
    }

    @Override
    public int getHearRadius()
    {
        // slightly better than LOS, and date of time does not affect our senses
        return (int) (7 + 1.5 * fStats.Per);
    }

    @Override
    public int getDamageType()
    {
        int dmgTypeId = Damage.DMG_GENERIC;

        for (int i = 0; i < this.fInventory.getCount(); i++) {
            Item item = this.fInventory.getItem(i);

            if (item.isUsed()) {
                int dmgType = item.getDamageType();

                if (dmgType != Damage.DMG_GENERIC) {
                    dmgTypeId = dmgType;
                }
            }
        }

        return dmgTypeId;
    }

    @Override
    public int getEquipBonus(String bonusId)
    {
        int bonus = 0;

        for (int i = 0; i < this.fInventory.getCount(); i++) {
            Item item = this.fInventory.getItem(i);

            if (item.isUsed()) {
                
            }
        }

        return bonus;
    }

    @Override
    public int getDamageValue()
    {
        return this.fStats.Str + this.getEquipBonus("damage");
    }

    public int getDefenceValue()
    {
        return this.getEquipBonus("defence");
    }

    @Override
    public void takeDamage(Damage damage)
    {
        super.takeDamage(damage);
        this.getSpace().addMessage(this.getName() + " took " + damage.Value + " damage", new Color(255, 0, 255));

        HumanBody body = (HumanBody) this.getBody();
        if (body != null) {
            body.takeDamage(damage);
        }

        if (!this.isAlive()) {
            Creature inflictor = damage.getInflictor();
            inflictor.kill(this);
        }
    }

    @Override
    public void attack(Creature enemy)
    {
        this.getSpace().addMessage(this.getName() + " is attacking " + enemy.getName(), new Color(255, 128, 128));

        CriminalEvent event = new CriminalEvent(this.getLocation(), this);
        event.post();

        if (this.isPlayer()) {
            SuspiciousSoundEvent soundEvent = new SuspiciousSoundEvent(this.getLocation(), 15);
            soundEvent.post();
        }
    }

    @Override
    public void die(Creature killer)
    {
        super.die(killer);

        if (this.isPlayer()) {
            this.getSpace().addMessage("You were killed by a " + killer.getName() + ".");
        } else {
            this.getSpace().addMessage(this.getName() + " has died.", Color.red);
        }
    }

    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="IEquippedCreature implementation">

    @Override
    public final int getMoney()
    {
        int result = 0;

        /*int num = this.fItems.getCount();
        for (int i = 0; i < num; i++) {
            Item item = this.fItems.getItem(i);
            if (item.CLSID == GlobalVars.iid_Coin) {
                result += (int) item.Count;
            }
        }*/

        return result;
    }

    // </editor-fold>
}

File: MysteriesRL/project/Creatures/Human.cs
/*
 *  "MysteriesRL", roguelike game.
 *  Copyright (C) 2015, 2017 by Serg V. Zhdanovskih.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System.Collections.Generic;
using System.Text;
using BSLib;
using MysteriesRL.Game;
using MysteriesRL.Game.Events;
using MysteriesRL.Items;
using MysteriesRL.Maps.Buildings;
using ZRLib.Core;
using ZRLib.Core.Body;
using ZRLib.Engine;
using ZRLib.Map;

namespace MysteriesRL.Creatures
{
    public class Human : Creature
    {
        private int fAge = 30;
        private Building fApartment;
        private AbstractBody fBody;
        private readonly IList<Human> fChildrens;
        private readonly IList<CrimeRecord> fCrimeRecords;
        private string fName;
        private readonly NPCStats fStats;
        private readonly ItemsList fInventory;
        private Human fParent;
        private readonly Personality fPersonality;
        private CreatureSex fSex = CreatureSex.csMale;
        private readonly IList<Human> fSiblings;
        private Human fSpouse;


        public bool Adult
        {
            get {
                return fAge >= 18;
            }
        }

        public int Age
        {
            get {
                return fAge;
            }
            set {
                fAge = value;
            }
        }

        public IList<CrimeRecord> CrimeRecords
        {
            get {
                return fCrimeRecords;
            }
        }

        public Human Parent
        {
            get {
                return fParent;
            }
            set {
                fParent = value;
            }
        }

        public CreatureSex Sex
        {
            get {
                return fSex;
            }
            set {
                fSex = value;
            }
        }

        public Human Spouse
        {
            get {
                return fSpouse;
            }
            set {
                if (fSpouse == null) {
                    fSpouse = value;
                    value.Spouse = this;
                }
            }
        }

        public override int DamageValue
        {
            get {
                return fStats.Str + GetEquipBonus("damage");
            }
        }

        public virtual int DefenceValue
        {
            get {
                return GetEquipBonus("defence");
            }
        }

        public override int HPMax
        {
            get {
                return fStats.HPMax;
            }
        }

        public override int FovRadius
        {
            get {
                int maxFov = (int)(7 + 1.2 * fStats.Per);
                int minFov = (int)(maxFov * 0.7);
    
                int fov = (int)(minFov + (maxFov - minFov) * Space.LightFactor);
                return fov;
            }
        }

        public override int HearRadius
        {
            get {
                // slightly better than LOS, and date of time does not affect our senses
                return (int)(7 + 1.5 * fStats.Per);
            }
        }

        public override int DamageType
        {
            get {
                int dmgTypeId = Damage.DMG_GENERIC;
    
                for (int i = 0; i < fInventory.Count; i++) {
                    Item item = fInventory[i];
    
                    if (item.Used) {
                        int dmgType = item.DamageType;
    
                        if (dmgType != Damage.DMG_GENERIC) {
                            dmgTypeId = dmgType;
                        }
                    }
                }
    
                return dmgTypeId;
            }
        }

        public int Money
        {
            get {
                int result = 0;
    
                /*int num = fItems.getCount();
                for (int i = 0; i < num; i++) {
                    Item item = fItems.getItem(i);
                    if (item.CLSID == GlobalVars.iid_Coin) {
                        result += (int) item.Count;
                    }
                }*/
    
                return result;
            }
        }

        public override char Appearance
        {
            get {
                if (Player) {
                    return '@';
                }
    
                if (Alive) {
                    switch (fSex) {
                        case CreatureSex.csFemale:
                            if (Adult) {
                                return 'W';
                            } else {
                                return 'w';
                            }
    
                        case CreatureSex.csMale:
                            if (Adult) {
                                return 'M';
                            } else {
                                return 'm';
                            }
    
                        default:
                            return 'H';
                    }
                } else {
                    return '%';
                }
            }
        }

        public override int AppearanceColor
        {
            get {
                if (Player) {
                    return Colors.Yellow;
                }
    
                return Colors.White;
            }
        }

        public override string Name
        {
            get {
                return fName;
            }
            set {
                fName = value;
            }
        }

        public Building Apartment
        {
            get {
                return fApartment;
            }
            set {
                fApartment = value;
            }
        }

        public NPCStats Stats
        {
            get {
                return fStats;
            }
        }

        public Personality Personality
        {
            get {
                return fPersonality;
            }
        }

        public AbstractBody Body
        {
            get {
                return fBody;
            }
            set {
                if (fBody != null) {
                    fBody.Dispose();
                }
                fBody = value;
            }
        }

        public override string Desc
        {
            get {
                StringBuilder result = new StringBuilder();
                result.Append(Name + ". ");
    
                bool isMale = (Sex == CreatureSex.csMale);
    
                string age;
                if (isMale) {
                    age = Locale.Format(RS.Rs_HeIsNYearsOld, Age);
                } else {
                    age = Locale.Format(RS.Rs_SheIsNYearsOld, Age);
                }
                result.Append(age);
    
                /*if (getSpouse() != null) {
                    result.append("His/her spouse " + getSpouse().getName() + ". ");
                }
        
                if (getParent() != null) {
                    result.append("His/her parent " + getParent().getName() + ". ");
                }*/
    
                int rsid;
                if (isMale) {
                    rsid = RS.Rs_HeHasXHairAndYEyes;
                } else {
                    rsid = RS.Rs_SheHasXHairAndYEyes;
                }
                string pers = Locale.Format(rsid, fPersonality.HairColor.ToString().ToLower(), fPersonality.EyeColor.ToString().ToLower());
                result.Append(pers);
    
                return result.ToString();
            }
        }

        public ItemsList Inventory
        {
            get {
                return fInventory;
            }
        }


        public Human(GameSpace space, object owner)
            : base(space, owner)
        {
            fBody = new HumanBody(this);
            fStats = new NPCStats();
            fInventory = new ItemsList(this);
            fPersonality = new Personality();

            fParent = null;
            fSpouse = null;
            fChildrens = new List<Human>();
            fSiblings = new List<Human>();
            fCrimeRecords = new List<CrimeRecord>();

            HP = HPMax;
        }

        public override void DoTurn()
        {
            HumanBody body = (HumanBody)fBody;
            body.Update();

            if (!body.Stunned) {
                base.DoTurn();
            }
        }

        public override void MoveTo(int newX, int newY)
        {
            HumanBody body = (HumanBody)fBody;
            if (body.Fainted) {
                return;
            }

            base.MoveTo(newX, newY);

            body.AdjustAttribute("stamina", -0.5f);
        }

        public virtual void Ascend(ExtPoint target)
        {
            IMap map = Map;
            if (map == Space.CurrentRealm.PlainMap) {
                //
            }
            if (map == Space.CurrentRealm.UndergroundMap) {
                Map = Space.CurrentRealm.PlainMap;
                MoveTo(target.X, target.Y);
            }
        }

        public virtual void Descend(ExtPoint target)
        {
            IMap map = Map;
            if (map == Space.CurrentRealm.PlainMap) {
                Map = Space.CurrentRealm.UndergroundMap;
                MoveTo(target.X, target.Y);
            }
        }

        public void AddChild(Human child)
        {
            foreach (Human otherChild in fChildrens) {
                otherChild.AddSibling(child);
            }

            fChildrens.Add(child);
            child.Parent = this;
        }

        public void AddSibling(Human child)
        {
            if (!fSiblings.Contains(child)) {
                fSiblings.Add(child);
                child.AddSibling(this);
            }
        }

        public void AddCrimeRecord(CrimeRecord crimeRecord)
        {
            foreach (CrimeRecord crime in fCrimeRecords) {
                if (crime.Type == crimeRecord.Type) {
                    crime.IncCount();
                    return;
                }
            }
            fCrimeRecords.Add(crimeRecord);
        }

        public override int GetEquipBonus(string bonusId)
        {
            int bonus = 0;

            for (int i = 0; i < fInventory.Count; i++) {
                Item item = fInventory[i];

                if (item.Used) {

                }
            }

            return bonus;
        }

        public override void TakeDamage(Damage damage)
        {
            base.TakeDamage(damage);
            Space.AddMessage(Name + " took " + damage.Value + " damage", Colors.Fuchsia);

            HumanBody body = (HumanBody)Body;
            if (body != null) {
                body.TakeDamage(damage);
            }

            if (!Alive) {
                Creature inflictor = damage.Inflictor;
                inflictor.Kill(this);
            }
        }

        protected override void Attack(Creature enemy)
        {
            Space.AddMessage(Name + " is attacking " + enemy.Name, 0xFF8080);

            CriminalEvent @event = new CriminalEvent(Location, this);
            @event.Post();

            if (Player) {
                SuspiciousSoundEvent soundEvent = new SuspiciousSoundEvent(Location, 15);
                soundEvent.Post();
            }
        }

        public override void Die(Creature killer)
        {
            base.Die(killer);

            if (Player) {
                Space.AddMessage("You were killed by a " + killer.Name + ".");
            } else {
                Space.AddMessage(Name + " has died.", Colors.Red);
            }
        }
    }
}
